### PR TITLE
Add (refactor) configuration contribution, allowing plugins to declare a schema for their configurables

### PR DESCRIPTION
**Pull request by tlambert03**
This revamps our configuration contribution.  It follows vscode's [`contributes.configuration`](https://code.visualstudio.com/api/references/contribution-points#contributes.configuration) spec pretty closely.

Plugins can register a schema (in [JSON schema](https://json-schema.org) format) for their "configurable".  This schema will be used both to provide a UI for the user to modify settings in napari, as well as to validate any stored settings.  The schema is a superset of [JSON schema draft 7](https://json-schema.org/specification-links.html#draft-7), with notable exceptions of `$ref` and `$def`, since we want plugin settings to be entirely self-contained.

It may appear rather verbose to define the Draft 7 schema in a pydantic model here, but I think the abstraction is worth it for the additional "explicitness" and flexibility, and it allows us to more easily autogenerate documentation.  An alternative here would be to have settings provide some `python_path` to a pydantic model in their module somewhere: but that a) ties us too tightly to pydantic, and b) leaves it way to open to user-error and or type incompatibilities if they were to start using custom types.

This PR doesn't actually implement the storage of settings, or any UI representation, but it does add an "extra" dependency on jsonschema (which napari would use) to validate an instance (object) against the plugin's provided schema.

(More manually written documentation would accompany this )

### CODE REVIEW

**Comment by tlambert03** [comment]
@nclack, this is ready to go.  It's very similar to what was already in `contributions/_configuration.py` but I revamped the JSON schema model, added a bunch of tests, and "unignored" the `configuration` contribution from the coverage requirements

File: tests/sample/my_plugin/napari.yaml [hunk 1]
```diff
@@ -29,13 +29,14 @@ contributions:
...
       python_name: my_plugin:make_widget_from_function
-  configuration: # call it settings?
-    properties:
-      my_plugin.reader.lazy:
-        type: boolean
-        default: false
-        title: Load lazily
-        description: Whether to load images lazily with dask
+  configuration:
+    - title: My Plugin
```
Comment (nclack) [inline tests/sample/my_plugin/napari.yaml:33]: ❤️ these examples

**Review by nclack — APPROVED** [review]
Super excited for this.  

I tried to leave a lot of suggestions on the schema, but I think it's fine as is.

File: npe2/manifest/contributions/_configuration.py [hunk 2]
```diff
@@ -1,116 +1,68 @@
...
+        "field, except for booleans, where the description is used as the label for "
+        "the checkbox. See also `markdown_description`.",
     )
-    writeOnly: Optional[bool]
-    properties: Optional[Dict[str, "JsonSchemaObject"]]
-    required: List[str] = []
-    nullable: Optional[bool] = False
-    x_enum_varnames: List[str] = Field(default_factory=list)
-    description: Optional[str]
     markdown_description: Optional[str] = Field(
```
[the remark this answers is not in the document]
Comment (tlambert03) [reply npe2/manifest/contributions/_configuration.py:41]: > I guess I'd rather have something like description_format where plain and markdown are options.

love it. how about one description, where the _default_ is markdown (of which plain text is of course a subset), but they can opt out of parsing with `plain`.

> I'm not a big fan of people having to write more than one description. I'm tempted to say markup support isn't worth the complexity.

definitely agree with this.  it was always going to be an "either/or" thing.  If they gave markdown, it overrides the regular.  So I like having a single definition field with format option much better.  (and if we come across a place where having markdown is somehow bad, I'm sure we can strip it if we must).

File: npe2/manifest/contributions/_configuration.py [hunk 3]
```diff
@@ -123,139 +75,85 @@ class JsonSchemaObject(BaseModel):
...
-    def validate_exclusive_maximum_and_exclusive_minimum(
-        cls, values: Dict[str, Any]
-    ) -> Any:
-        if "$ref" in values:
-            # ref not supported
-            values.pop("$ref")
-
-        exclusive_maximum: Union[float, bool, None] = values.get("exclusiveMaximum")
-        exclusive_minimum: Union[float, bool, None] = values.get("exclusiveMinimum")
+    scope: Literal["application"] = Field(
```
[the remark this answers is not in the document]
Comment (tlambert03) [reply npe2/manifest/contributions/_configuration.py:79]: yeah, I struggled with this one too. let's leave it out for now.
I can imagine a day when scope becomes relevant (i.e. where a user might want to be able to configure a plugin one way in one [workspace](https://github.com/napari/napari/issues/4227), and another way in another workspace), but that day is _far_ away.  So let's YAGNI this until we have any idea about what actually supporting it means.

File: npe2/manifest/contributions/_configuration.py [hunk 4]
```diff
@@ -123,139 +75,85 @@ class JsonSchemaObject(BaseModel):
...
+        description="A configuration setting can have one of the following scopes:\n"
+        "  - `application`: Applies globally, can only be configured in user settings",
+        # TODO: add support for these additional scopes
+        # "  - `machine`: Machine specific settings that can be set only in user "
+        # "settings or only in remote settings. For example, an installation path which"
+        # " shouldn't be shared across machines.\n"
+        # "  - `window`: Viewer (instance) specific settings which can be configured"
+        # "in user or workspace settings.",  # this should be the default
+    )
+    edit_presentation: Literal["singleline", "multiline"] = Field(
```
[the remark this answers is not in the document]
Comment (tlambert03) [reply npe2/manifest/contributions/_configuration.py:90]: that one is more of a hint for what the input widget would be (i.e. a `QLineEdit` vs a `QTextEdit`)... indicating that you want the user to have a multi-line text editor when they're presented this field.  but maybe I'm missing the point? and this _is_ rather "string" specific...  can you guide me with an example?

File: npe2/manifest/contributions/_configuration.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nclack) [reply npe2/manifest/contributions/_configuration.py]: yay

File: npe2/manifest/contributions/_configuration.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nclack) [reply npe2/manifest/contributions/_configuration.py:78]: Something like

```python
class Description: # or Text or something
  text: str # <-- the data
  # Styling/presentation parameters below: 
  markup: Literal['plain','markdown']
  edit_presentation: Literal['singleline','multiline']
  
class ConfigurationProperty:
   ...
   description: str | Description
   ... 
```

As an aside, I'm not convinced 'edit_presentation' is necessary - naively I'd say just infer that from a newline present in the text.

[the remark this answers is not in the document]
Comment (tlambert03) [reply npe2/manifest/contributions/_configuration.py:78]: But there isn't necessarily any text yet.  This is user input, (not plugin provided), and there may not be a "default" value to infer from

[the remark this answers is not in the document]
Comment (tlambert03) [reply npe2/manifest/contributions/_configuration.py:78]: Edit_presentation isn't related to "description", it's related more to "type" and only relevant when "type: string"

[the remark this answers is not in the document]
Comment (nclack) [reply npe2/manifest/contributions/_configuration.py:78]: ah ha! I missed that completely. Thanks. Will resolve this thread.

**Pull request merged**